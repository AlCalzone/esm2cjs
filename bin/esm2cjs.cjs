#!/usr/bin/env node

const yargs = require("yargs/yargs");
const { hideBin } = require("yargs/helpers");
const path = require("path");
const { esm2cjs } = require("../build/cjs");

const argv = yargs(hideBin(process.argv))
	.option("in", {
		type: "string",
		demandOption: true,
		description: "The input directory containing ESM sources",
	})
	.option("out", {
		type: "string",
		demandOption: true,
		description:
			"The output directory where CommonJS files should be placed",
	})
	.option("glob", {
		alias: "g",
		array: true,
		type: "string",
		default: "**/*.js",
		description: "Override the default input file glob(s)",
	})
	.option("platform", {
		alias: "p",
		type: "string",
		default: "node",
		choices: ["node", "browser", "neutral"],
		description:
			"Change the platform the code is meant to be run on. Details see https://esbuild.github.io/api/#platform",
	})
	.option("target", {
		alias: "t",
		type: "string",
		default: "node18",
		description:
			"Override the default compile target. Details see https://esbuild.github.io/api/#target",
	})
	.option("loglevel", {
		alias: "l",
		type: "string",
		choices: ["verbose", "debug", "info", "warning", "error", "silent"],
		default: "warning",
	})
	.option("sourcemap", {
		alias: "m",
		type: "boolean",
		default: true,
		description:
			"The output directory where CommonJS files should be placed",
	})
	.option("writePackageJson", {
		alias: "j",
		type: "boolean",
		default: true,
		description:
			"Create two package.json files containing the module type in the input and output directory",
	})
	.option("packageJsonSideEffects", {
		alias: "x",
		type: "boolean",
		default: true,
		description:
			"When writePackageJson is true, this can be used to specify whether the generated package.json files should contain sideEffects: false. Bundlers use this information to tree-shake the code. " +
			"To specify individual files with side effects instead, use the programmatic API.",
	})
	.option("cleanOutDir", {
		alias: "c",
		type: "boolean",
		default: false,
		description: "Empty the output directory before writing",
	})
	.example(
		"$0 --in test/esm --out test/cjs",
		"Compiles .js files in test/esm to CommonJS and places them in test/cjs",
	)
	.parseSync();

(async () => {
	const inDir = path.join(process.cwd(), argv.in);
	const outDir = path.join(process.cwd(), argv.out);

	await esm2cjs({
		inDir,
		outDir,
		cleanOutDir: argv.cleanOutDir,
		globs: argv.glob,
		target: argv.target,
		platform: /** @type {any} */ (argv.platform),
		writePackageJson: argv.writePackageJson,
		logLevel: /** @type {any} */ (argv.loglevel),
	});
})().catch((e) => {
	console.error(e);
	process.exit(1);
});
