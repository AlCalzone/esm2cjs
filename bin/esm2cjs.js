#!/usr/bin/env node

const yargs = require("yargs/yargs");
const { hideBin } = require("yargs/helpers");
const path = require("path");
const { esm2cjs } = require("..");

const argv = yargs(hideBin(process.argv))
	.option("in", {
		type: "string",
		demandOption: true,
		description: "The input directory containing ESM sources",
	})
	.option("out", {
		type: "string",
		demandOption: true,
		description:
			"The output directory where CommonJS files should be placed",
	})
	.option("glob", {
		alias: "g",
		array: true,
		type: "string",
		default: "**/*.js",
		description: "Override the default input file glob(s)",
	})
	.option("loglevel", {
		alias: "l",
		type: "string",
		choices: ["verbose", "debug", "info", "warning", "error", "silent"],
		default: "warning",
	})
	.option("sourcemap", {
		alias: "m",
		type: "boolean",
		default: true,
		description:
			"The output directory where CommonJS files should be placed",
	})
	.option("writePackageJson", {
		alias: "p",
		type: "boolean",
		default: true,
		description:
			"Create two package.json files containing the module type in the input and output directory",
	})
	.option("cleanOutDir", {
		alias: "c",
		type: "boolean",
		default: false,
		description: "Empty the output directory before writing",
	})
	.example(
		"$0 --in test/esm --out test/cjs",
		"Compiles .js files in test/esm to CommonJS and places them in test/cjs",
	).argv;

(async () => {
	const inDir = path.join(process.cwd(), argv.in);
	const outDir = path.join(process.cwd(), argv.out);

	await esm2cjs({
		inDir,
		outDir,
		cleanOutDir: argv.cleanOutDir,
		globs: argv.glob,
		writePackageJson: argv.writePackageJson,
		logLevel: /** @type {any} */ (argv.loglevel),
	});
})().catch((e) => {
	console.error(e);
	process.exit(1);
});
