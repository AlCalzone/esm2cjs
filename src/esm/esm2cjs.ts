import path from "node:path";
import { BuildOptions, build } from "esbuild";
import fs from "fs-extra";
import glob from "tiny-glob";
import { fileURLToPath } from "node:url";

const _dirname = path.dirname(fileURLToPath(import.meta.url));
export const shimsDir = path.join(_dirname, "../../shims");

export interface ESM2CJSOptions {
	inDir: string;
	outDir: string;
	cleanOutDir?: boolean;
	globs?: string | string[];
	sourcemap?: boolean;
	logLevel?: BuildOptions["logLevel"];
	platform?: BuildOptions["platform"];
	target?: BuildOptions["target"];
	writePackageJson?: boolean;
}

export async function esm2cjs({
	inDir,
	outDir,
	globs = ["**/*.js"],
	sourcemap = true,
	logLevel = "warning",
	platform = "node",
	target = "node18",
	cleanOutDir = false,
	writePackageJson = true,
}: ESM2CJSOptions) {
	// Clean the output dir if necessary
	if (cleanOutDir) await fs.emptyDir(outDir);

	// Figure out what to compile
	if (typeof globs === "string") globs = [globs];
	const entryPoints = (
		await Promise.all(globs.map((g) => glob(g, { cwd: inDir })))
	).reduce((prev, cur) => [...prev, ...cur], []);

	// Compile ESM to CJS using esbuild
	const buildResult = await build({
		absWorkingDir: inDir,
		entryPoints,
		outdir: outDir,
		bundle: false,
		minify: false,
		metafile: true,
		sourcemap,
		logLevel,
		platform,
		format: "cjs",
		target,
		define: {
			"import.meta.url": "__import_meta_url",
		},
		inject: [path.join(shimsDir, "import.meta.url/shim.js")],
	});

	// Copy .d.ts files
	if (buildResult.metafile) {
		for (const inputFile of Object.keys(buildResult.metafile.inputs)) {
			if (inputFile.startsWith(".")) {
				// File outside of the build directory, skip
				continue;
			}
			const declarationFileName = inputFile.replace(
				/\.([cm]?)js$/,
				".d.$1ts",
			);
			// Try to copy the file, but don't fail if it doesn't exist
			await fs
				.copyFile(
					path.join(inDir, declarationFileName),
					path.join(outDir, declarationFileName),
				)
				.catch(() => {});
		}
	}

	// If desired, define the module type of each build directory separately
	if (writePackageJson) {
		await fs.writeJSON(
			path.join(inDir, "package.json"),
			{ type: "module" },
			{
				spaces: 4,
			},
		);
		await fs.writeJSON(
			path.join(outDir, "package.json"),
			{ type: "commonjs" },
			{
				spaces: 4,
			},
		);
	}
}
