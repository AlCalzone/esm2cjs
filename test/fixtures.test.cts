import { esm2cjs } from "../build/cjs/esm2cjs.js";
import fs from "fs-extra";
import path from "path";
import {describe, beforeAll, it, expect} from "vitest"

describe("Compilation tests", () => {
	beforeAll(async () => {
		await esm2cjs({
			inDir: path.join(__dirname, "fixtures/esm"),
			outDir: path.join(__dirname, "fixtures/cjs"),
			cleanOutDir: true,
			writePackageJson: true,
		});
	});

	it("Test 1: .js", async () => {
		const content = await fs.readFile(
			path.join(__dirname, "fixtures/cjs/test1.js"),
			"utf8",
		);

		expect(content).toMatchSnapshot();

		expect(require("./fixtures/cjs/test1.js").test).toBe("test1");
	});

	it("Test 2: import.meta.url", async () => {
		const content = await fs.readFile(
			path.join(__dirname, "fixtures/cjs/test2.js"),
			"utf8",
		);

		expect(content).toMatchSnapshot();

		expect(require("./fixtures/cjs/test2.js")._dirname).toBe(
			path.join(__dirname, "fixtures/cjs"),
		);
	});

	it("Test 3: package #imports", async () => {
		const content = await fs.readFile(
			path.join(__dirname, "fixtures/cjs/test3.js"),
			"utf8",
		);

		expect(content).toMatchSnapshot();

		const esmEnv = await import("./fixtures/esm/test3.js");
		const cjsEnv = require("./fixtures/cjs/test3.js");
		
		expect(esmEnv.env).toEqual("esm");
		expect(cjsEnv.env).toEqual("cjs");
	});
});
