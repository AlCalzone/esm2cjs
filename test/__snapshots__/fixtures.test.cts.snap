// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compilation tests > Test 1: .js 1`] = `
""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var test1_exports = {};
__export(test1_exports, {
  test: () => test
});
module.exports = __toCommonJS(test1_exports);
const test = "test1";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  test
});
//# sourceMappingURL=test1.js.map
"
`;

exports[`Compilation tests > Test 2: import.meta.url 1`] = `
""use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var test2_exports = {};
__export(test2_exports, {
  _dirname: () => _dirname
});
module.exports = __toCommonJS(test2_exports);
var __import_meta_url = typeof document === "undefined" ? new (require("url".replace("", ""))).URL("file:" + __filename).href : document.currentScript && document.currentScript.src || new URL("main.js", document.baseURI).href;
var import_node_path = __toESM(require("node:path"), 1);
var import_node_url = require("node:url");
const _dirname = import_node_path.default.dirname((0, import_node_url.fileURLToPath)(__import_meta_url));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  _dirname
});
//# sourceMappingURL=test2.js.map
"
`;
